src/SmartFormat/Extensions/DictionarySource.cs

```diff
@@ -25,28 +33,119 @@ public override bool TryEvaluateSelector(ISelectorInfo selectorInfo)
         if (current is IDictionary rawDict)
             foreach (DictionaryEntry entry in rawDict)
             {
                 var key = entry.Key as string ?? entry.Key.ToString()!;
 
-                if (key.Equals(selector, selectorInfo.FormatDetails.Settings.GetCaseSensitivityComparison()))
-                {
-                    selectorInfo.Result = entry.Value;
-                    return true;
-                }
+                if (!key.Equals(selector, selectorInfo.FormatDetails.Settings.GetCaseSensitivityComparison()))
+                    continue;
+
+                selectorInfo.Result = entry.Value;
+                return true;
             }
 
-        // this check is for dynamics and generic dictionaries
-        if (current is not IDictionary<string, object?> dict) return false;
+        // This check is for dynamics (ExpandoObject):
+        if (current is IDictionary<string, object?> dict)
+        {
+            // We're using the CaseSensitivityType of the dictionary,
+            // not the one from Settings.GetCaseSensitivityComparison().
+            // This is faster and has less GC than Key.Equals(...)
+            if (!dict.TryGetValue(selector, out var val)) return false;
+
+            selectorInfo.Result = val;
+            return true;
+        }
+        
+        // This is for IReadOnlyDictionary<,> using Reflection
+        if (IsIReadOnlyDictionaryEnabled && TryGetDictionaryValue(current, selector,
+                selectorInfo.FormatDetails.Settings.GetCaseSensitivityComparison(), out var value))
+        {
+            selectorInfo.Result = value;
+            return true;
+        }
+        
+        return false;
+    }
+
+    #region *** IReadOnlyDictionary<,> ***
+
+    /// <summary>
+    /// Gets the type cache <see cref="IDictionary{TKey,TValue}"/> for <see cref="IReadOnlyDictionary{TKey,TValue}"/>.
+    /// It could e.g. be pre-filled or cleared in a derived class.
+    /// </summary>
+    /// <remarks>
+    /// Note: For reading, <see cref="Dictionary{TKey, TValue}"/> and <see cref="ConcurrentDictionary{TKey,TValue}"/> perform equally.
+    /// For writing, <see cref="ConcurrentDictionary{TKey, TValue}"/> is slower with more garbage (tested under net5.0).
+    /// </remarks>
+    protected internal readonly IDictionary<Type, (PropertyInfo, PropertyInfo)?> RoDictionaryTypeCache =
+        SmartSettings.IsThreadSafeMode
+            ? new ConcurrentDictionary<Type, (PropertyInfo, PropertyInfo)?>()
+            : new Dictionary<Type, (PropertyInfo, PropertyInfo)?>();
+
+    /// <summary>
+    /// Gets or sets, whether the <see cref="IReadOnlyDictionary{TKey,TValue}"/> interface should be supported.
+    /// Although caching is used, this is still slower than the other types.
+    /// Default is <see langword="false"/>.
+    /// </summary>
+    public bool IsIReadOnlyDictionaryEnabled { get; set; } = false;
```

Comment:
Maybe name it something like `ReadOnlyDictionarySupportEnabled`

Reply:
Yes, `IsIReadOnlyDictionarySupportEnabled` is better indeed. Will change.

Reply:
... or shorter `IsIReadOnlyDictionarySupported` 😉